substrate/frame/staking/src/asset.rs

```diff
@@ -41,42 +49,78 @@ pub fn total_balance<T: Config>(who: &T::AccountId) -> BalanceOf<T> {
+	T::Currency::balance_on_hold(&HoldReason::Staking.into(), who)
+}
+
+/// Balance of who that can be staked additionally.
+///
+/// Does not include the current stake.
+pub fn free_to_stake<T: Config>(who: &T::AccountId) -> BalanceOf<T> {
+	T::Currency::balance(who)
 }
 
 /// Set balance that can be staked for `who`.
 ///
-/// This includes any balance that is already staked.
+/// `Value` must be greater than already staked plus existential deposit for free balance.
+///
+/// Should only be used with test.
 #[cfg(any(test, feature = "runtime-benchmarks"))]
 pub fn set_stakeable_balance<T: Config>(who: &T::AccountId, value: BalanceOf<T>) {
-	T::Currency::make_free_balance_be(who, value);
+	use frame_support::traits::fungible::Mutate;
+
+	let staked_balance = staked::<T>(who);
+	// if value is greater than staked balance, we need to increase the free balance.
+	if value > staked_balance {
+		let _ = T::Currency::set_balance(who, value - staked_balance);
+	} else {
+		// else reduce the staked balance.
+		update_stake::<T>(who, value).expect("can remove from what is staked");
+		// burn all free
+		let _ = T::Currency::set_balance(who, Zero::zero());
+	}
+
+	// ensure new stakeable balance same as desired `value`.
+	assert_eq!(stakeable_balance::<T>(who), value);
 }
 
 /// Update `amount` at stake for `who`.
 ///
 /// Overwrites the existing stake amount. If passed amount is lower than the existing stake, the
 /// difference is unlocked.
-pub fn update_stake<T: Config>(who: &T::AccountId, amount: BalanceOf<T>) {
-	T::Currency::set_lock(
-		crate::STAKING_ID,
-		who,
-		amount,
-		frame_support::traits::WithdrawReasons::all(),
-	);
+pub fn update_stake<T: Config>(who: &T::AccountId, amount: BalanceOf<T>) -> DispatchResult {
+	// if first stake, inc provider. This allows us to stake all free balance.
+	if staked::<T>(who) == Zero::zero() && amount > Zero::zero() {
+		frame_system::Pallet::<T>::inc_providers(who);
+	}
+
+	T::Currency::set_on_hold(&HoldReason::Staking.into(), who, amount)
 }
 
-/// Kill the stake of `who`.
+/// Release all staked amount to `who`.
 ///
-/// All locked amount is unlocked.
-pub fn kill_stake<T: Config>(who: &T::AccountId) {
-	T::Currency::remove_lock(crate::STAKING_ID, who);
+/// Fails if there are consumers left on `who` that restricts it from being reaped.
+pub fn kill_stake<T: Config>(who: &T::AccountId) -> DispatchResult {
+	T::Currency::release_all(&HoldReason::Staking.into(), who, Precision::BestEffort)
+		.map(|_| ())?;
+
+	if !frame_system::Pallet::<T>::can_dec_provider(who) {
+		// The session pallet keeps a consumer reference to validator stash which may block this
+		// pallet from clearing its provider reference. If the account cannot provide for itself
+		// (via enough free balance) to keep the session key, we try to clean up the session keys.
+		// If the extra consumer comes from somewhere else, reap stash would fail.
+		T::SessionInterface::purge_keys(who.clone())?;
```

Comment:
isn't storage dirty if an error is returned now? maybe we can just wrap into transactional to make it safe?

Reply:
We can go this way but I am pretty sure few people are looking at the rustdoc when reviewing a PR.
And people can easily miss it when changing code.

I am afraid it will backfire some day.